spaceship/lib/spaceship/tunes/tunes_client.rb

```diff
@@ -168,6 +176,100 @@ def send_login_request(user, password)
       end
     end
 
+    def handle_two_step(response)
+      @x_apple_web_session_token = response["x-apple-web-session-token"]
+      @scnt = response["scnt"]
+
+      r = request(:get) do |req|
+        req.url "https://idmsa.apple.com/appleauth/auth"
+        req.headers["scnt"] = @scnt
+        req.headers["X-Apple-Web-Session-Token"] = @x_apple_web_session_token
+        req.headers["Accept"] = "application/json"
+      end
+
+      if r.body.kind_of?(Hash) && r.body["trustedDevices"].kind_of?(Array)
+        if r.body.fetch("securityCode", {})["tooManyCodesLock"].to_s.length > 0
+          raise ITunesConnectError.new, "Too many verification codes have been sent. Enter the last code you received, use one of your devices, or try again later."
+        end
+
+        old_client = (begin
+                        Tunes::RecoveryDevice.client
+                      rescue
+                        nil # since client might be nil, which raises an exception
+                      end)
+        Tunes::RecoveryDevice.client = self # temporary set it as it's required by the factory method
+        devices = r.body["trustedDevices"].collect do |current|
+          Tunes::RecoveryDevice.factory(current)
+        end
+        Tunes::RecoveryDevice.client = old_client
+
+        puts "Two Step Verification for account '#{self.user}' is enabled"
+        puts "Please select a device to verify your identity"
+        available = devices.collect do |c|
+          "#{c.name}\t#{c.model_name || 'SMS'}\t(#{c.device_id})"
+        end
+        result = choose(*available)
+        device_id = result.match(/.*\t.*\t\((.*)\)/)[1]
+        select_device(r, device_id)
+      else
+        raise "Invalid 2 step response #{r.body}"
+      end
+    end
+
+    def select_device(r, device_id)
+      # Request Token
+      r = request(:put) do |req|
+        req.url "https://idmsa.apple.com/appleauth/auth/verify/device/#{device_id}/securitycode"
+        req.headers["Accept"] = "application/json"
+        req.headers["scnt"] = @scnt
+        req.headers["X-Apple-Web-Session-Token"] = @x_apple_web_session_token
+      end
+      handle_itc_response(r.body)
+
+      puts "Successfully requested notification"
+      code = ask("Please enter the 4 digit code: ")
+      puts "Requesting session..."
+
+      # Send token back to server to get a valid session
+      r = request(:post) do |req|
+        req.url "https://idmsa.apple.com/appleauth/auth/verify/device/#{device_id}/securitycode"
+        req.headers["Accept"] = "application/json"
+        req.headers["scnt"] = @scnt
+        req.headers["X-Apple-Web-Session-Token"] = @x_apple_web_session_token
+        req.body = { "code" => code.to_s }.to_json
+        req.headers['Content-Type'] = 'application/json'
+      end
+      handle_itc_response(r.body) # this will fail if the code is invalid
+      # If it works, r.body is actually nil
+      raise "Could not login" unless r.headers["set-cookie"].include?("myacinfo")
```

Comment:
Is there another way to determine that the response was successful? It seems brittle to go after a specific cookie key. I've noticed that iTC will respond with semantic HTTP response codes given a proper request. Maybe we check
```ruby
raise "Could not login"  if response.status == 401
```